### ISSUE TITLE
Tables: $unset does not always behave as $set to null in updateOne

### ISSUE THREAD

**Issue by hemidactylus** [stargate/data-api#1673]
UpdateOne. $setting to null is != $unsetting, when it's all non-key columns. (likely not a desired fact)

Assume table with columns: 'pk' and 'col2', 'col3' (pk the only primary key column)

## Set some fields to null

i.e. `{"filter": {"pk": "A"}, "update": {"$set": {"col2": null}}}`

- if row exists, puts those nulls in the row
- if does not exists, does nothing

## Set ALL (non-pk) fields to null

i.e. `{"filter": {"pk": "A"}, "update": {"$set": {"col2": null, "col3": null}}}`

- if row exists, *it is deleted from DB*
- if not exists, does nothing

## Unset some fields

i.e. `{"filter": {"pk": "A"}, "update": {"$unset": {"col2": ""}}}`

- if row exists, puts those nulls in the row
- if not exists, does nothing

## Unset ALL (non-pk) fields

i.e. `{"filter": {"pk": "A"}, "update": {"$unset": {"col2": "", "col3": ""}}}`

- if row exists, *puts all nulls in the row*
- if not exists, does nothing

# So ...

`$unset: <all fields>` is not the same as `$set: <all fields to null>`

**Comment by Yuqi-Du** [comment]
```
Set ALL (non-pk) fields to null
i.e. {"filter": {"pk": "A"}, "update": {"$set": {"col2": null, "col3": null}}}
if row exists, it is deleted from DB
```
I can not reproduce this one, just to confirm, are you saying the row is just getting dropped by doing this?

**Comment by hemidactylus** [comment]
> the row is just getting dropped by doing this? (i.e. `"update": {"$set": {"all non PK columns": null}}`)

Correct, this is what I am saying.

**Comment by Yuqi-Du** [comment]
Sorry, Stefano, could you please reproduce again, I can not reproduce it.
In theory, update statement shouldn't drop a row.

**Comment by hemidactylus** [comment]
Re-opening this. I uncovered the real reason for the surprising behaviour and will report here in a moment.

**Comment by hemidactylus** [comment]
# A CQL difference not surfaced immediately

Foreword: `$set: {f: null}` and `{$unset: {f: ""}}` are translated to the very same, identical CQL statement, so this cannot bear any difference. The reason for a surprising behaviour lies elsewhere. Read on for the details.

TL;DR = there is a hidden difference at the CQL level that the Data API faithfully brings to the user. Below a CQL reproducible example, and its Data API reformulation.

## CQL

Despite looking the same (e.g. in a SELECT), a row inserted with INSERT and one inserted with UPDATE have something different, which becomes apparent when unsetting (setting to null) all the non-PK columns. Create a table:

```
CREATE TABLE default_keyspace.table_upd (
    pk text PRIMARY KEY,
    col2 text,
    col3 text
);
```

and populate it with one row **inserted**,

```
insert into table_upd (pk, col2, col3) VALUES ( 'inserted', 'x', 'y');
```

and one created through an **update** statement:

```
update table_upd set col2='a', col3='b' where pk='born_by_update';
```

No visible difference e.g. if you `select *`:

```
 pk             | col2 | col3
----------------+------+------
       inserted |    x |    y
 born_by_update |    a |    b

(2 rows)
```

Now, run the _same_ statement on both rows, which sets to null the `col2` and `col3` fields. The result is very different (a surprise to me)

```
update table_upd set col2=null, col3=null where pk='inserted'; // has the effect that the row has pk+nulls on DB
update table_upd set col2=null, col3=null where pk='born_by_update';  // has the effect that the row disappears
```

As a result, `select *` now **lists only one row**:

```
 pk       | col2 | col3
----------+------+------
 inserted | null | null

(1 rows)
```

## Data API equivalent

Here are the payloads to copy paste for testing:

_Create_

```
{
    "createTable": {
        "name": "table_upd",
        "definition": {
            "columns": {
                "pk": {
                    "type": "text"
                },
                "col2": {
                    "type": "text"
                },
                "col3": {
                    "type": "text"
                }
            },
            "primaryKey": "pk"
        }
    }
}
```

_Insert_

```
{
    "insertOne": {
        "document": {
            "pk": "inserted", "col2": "x", "col3": "y"
        }
    }
}
```

_"Insert" by updating_

```
{
    "updateOne": {
        "filter": {
            "pk": "born_by_update"
        },
        "update": {
            "$set": {
                "col2": "a", "col3": "b"
            }
        }
    }
}
```

_Query the table_

```
{
    "find": {
        "filter": {}
    }
}
```

_Set non-PK columns to null (two payloads)_

```
{
    "updateOne": {
        "filter": {"pk": "inserted"},
        "update": {"$set": {"col2": null, "col3": null}}
    }
}

{
    "updateOne": {
        "filter": {"pk": "born_by_update"},
        "update": {"$set": {"col2": null, "col3": null}}
    }
}
```

**Comment by hemidactylus** [comment]
Well ... it turns out this is a known Cassandra issue - "known" but not widely. In practice only INSERTs create "row markers", which explains the different behaviour once setting all non-pk columns to null.

See this answer on SO: https://stackoverflow.com/a/60075479

I suppose this is not a Data API "bug" in itself, but worth perhaps a docs mention or something?

**Comment by aimurphy** [comment]
I think I'm having a hard time understanding the whole situation due to my naivety about CQL and Cassandra.

For documentation, there are some different approaches we can take (separately or a combination):

1. Try to enumerate edge case(s) where this unexpected delete can happen.

2. Link to the SO thread. I don't particularly want to link to a SO thread, but if someone *really* wants to know why it's happening, they have to dig into the underlying Cassandra/CQL stuff.

3. Ignore the reality and simply advise against certain actions (ie. don't use updateOne to unset an entire row, instead drop and re-insert it with only the PK values????)

These are just some ideas. I am open to others.

**Comment by amorton** [comment]
putting on hold, the actual bug is unclear, needs clearer expected and actual results.
